### PR TITLE
Retrieve genes transcripts and  exons using the schug library

### LINKED ISSUE #3856: Replace code that downloads genes, exons and transcripts from Ensembl with schug lib
We now have a dedicated repository to take care of these downloads: https://github.com/Clinical-Genomics/schug

We can replace big chunks of code in scout with just a few lines.

It is either downloading the files from a schug instance: https://schug-stage.scilifelab.se/docs or installing schug as a lib and using it inside scout (it is on PyPI).

I prefer the second option because this way all other facilities using scout would not be forced to deploy schug themselves or ask us to open our instance to them for updating those collections

### PR DESCRIPTION
**Pull request by northwestwitch**
- Remove many lines of code that build and send requests (via `utils.scout_requests`) for genes , transcripts and exons.
- Replace them with calls to the schug library (closes #3856)

<details>
<summary>Testing on cg-vm1 server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. Make sure the PR is pushed and available on [Docker Hub](https://hub.docker.com/repository/docker/clinicalgenomics/scout-server-stage)
1. Fist book your testing time using the Pax software available at [https://pax.scilifelab.se/](https://pax.scilifelab.se). The resource you are going to call dibs on is `scout-stage` and the server is `cg-vm1`.
1. `ssh <USER.NAME>@cg-vm1.scilifelab.se`
1. `sudo -iu hiseq.clinical`
1. `ssh localhost`
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `podman ps`
1. Stop the service with current deployed branch: `systemctl --user stop scout.target`
1. Start the scout service with the branch to test: `systemctl --user start scout@<this_branch>`
1. Make sure the branch is deployed: `systemctl --user status scout.target`
1. After testing is done, repeat procedure at [https://pax.scilifelab.se/](https://pax.scilifelab.se), which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

<details>
<summary>Testing on hasta server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. `ssh <USER.NAME>@hasta.scilifelab.se`
1. Book your testing time using the Pax software. `us; paxa -u <user> -s hasta -r scout-stage`. You can also use the WSGI Pax app available at [https://pax.scilifelab.se/](https://pax.scilifelab.se).
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `conda activate S_scout; pip freeze | grep scout-browser`
1. Deploy the branch to test: `bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_scout -t scout -b <this_branch>`
1. Make sure the branch is deployed: `us; scout --version`
1. After testing is done, repeat the `paxa` procedure, which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

**How to test**:
1. Locally, download the ensembl files with the following command: `scout --demo download ensembl --exons -o <dir-of-choice>`
2. Run `scout --demo update genes -f <dir-of-choice>`
3. Run `scout --demo update genes` and make sure the genes get downloaded and updated on the fly

**Expected outcome**:
The commands above should work fine

**Review:**
- [x] code approved by DN
- [x] tests executed by CR

### CODE REVIEW

**Comment by northwestwitch** [comment]
This is ready for review now

File: scout/commands/update/genes.py [hunk 1]
```diff
@@ -81,7 +81,7 @@ def fetch_downloaded_resources(resources, downloads_folder, builds):
             if resource_exists:
                 resources[resname] = get_file_handle(resource_path).readlines()
 
-        # If the resource is manadatory make sure it exists and contains data (OMIM data is NOT mandatory)
+        # If the resource is mandatory, make sure it exists and contains data (OMIM data is NOT mandatory)
```
Comment (dnil) [inline scout/commands/update/genes.py:84]: ✅

File: requirements.txt [hunk 2]
```diff
@@ -23,6 +23,7 @@ click
 cryptography
 defusedxml
 svglib
+schug
```
Comment (dnil) [inline requirements.txt:26]: :tada:

File: CHANGELOG.md [hunk 3]
```diff
@@ -38,6 +38,7 @@ About changelog [here](https://keepachangelog.com/en/1.0.0/)
 - Allow ClinVar submissions with custom API key for users registered as ClinVar submitters or when institute doesn't have a preset list of ClinVar submitters
 - Ordered event verbs alphabetically and created ClinVar-related user events
 - Removed the unused "no-variants" option from the load case command line
+- Replace the code that downloads Ensembl genes, transcripts and exons with schug library
```
Comment (dnil) [inline CHANGELOG.md:41]:
```suggestion
- Replace the code that downloads Ensembl genes, transcripts and exons with the Schug library
```

File: scout/commands/download/ensembl.py [hunk 4]
```diff
@@ -1,29 +1,20 @@
...
 
 LOG = logging.getLogger(__name__)
 
 
-def print_ensembl(out_dir, resource_type, genome_build=None):
-    """Fetch and print ensembl info to file
+def print_ensembl(
+    out_dir: pathlib.Path, resource_type: List[str], genome_build: Optional[str] = None
+):
+    """Fetch and print Ensembl info to file. If no genome build is used both builds will be fetched."""
```
Comment (dnil) [inline scout/commands/download/ensembl.py:16]: 🥇 
```suggestion
    """Fetch and print Ensembl info to file. If no genome build is passed as argument, both builds will be fetched."""
```

File: scout/commands/download/ensembl.py [hunk 5]
```diff
@@ -33,23 +24,20 @@ def print_ensembl(out_dir, resource_type, genome_build=None):
...
-    elif resource_type == "transcripts":
-        file_name = "ensembl_transcripts_{}.txt"
-        fetch_function = fetch_ensembl_transcripts
-    else:
-        file_name = "ensembl_exons_{}.txt"
-        fetch_function = fetch_ensembl_exons
 
     for build in builds:
-        file_path = out_dir / file_name.format(build)
+        ensembl_client = EnsemblBiomartHandler(build=build)
```
Comment (dnil) [inline scout/commands/download/ensembl.py:29]: 🧹💨

File: scout/load/setup.py [hunk 6]
```diff
@@ -134,13 +133,15 @@ def setup_scout(
...
+    # Load genes
     builds = ["37", "38"]
     for build in builds:
         genes_path = "genes{}_path".format(build)
+        ensembl_client = EnsemblBiomartHandler(build=build)
         if resource_files.get(genes_path):
             ensembl_genes = get_file_handle(resource_files[genes_path])
         else:
-            ensembl_genes = fetch_ensembl_genes(build=build)
+            ensembl_genes = ensembl_client.stream_resource(interval_type="genes")
```
Comment (dnil) [inline scout/load/setup.py:144]: 🥇

File: scout/load/setup.py [hunk 7]
```diff
@@ -134,13 +133,15 @@ def setup_scout(
     for genome_build, cytobands_path in cytoband_files.items():
         load_cytobands(cytobands_path, genome_build, adapter)
 
+    # Load genes
     builds = ["37", "38"]
     for build in builds:
         genes_path = "genes{}_path".format(build)
+        ensembl_client = EnsemblBiomartHandler(build=build)
```
Comment (dnil) [inline scout/load/setup.py:140]: ✅

File: scout/parse/ensembl.py [hunk 8]
```diff
@@ -5,9 +5,9 @@
 
 
 def parse_ensembl_line(line, header):
-    """Parse an ensembl formated line
+    """Parse an ensembl formatted line
```
Comment (dnil) [inline scout/parse/ensembl.py:8]: ✅

File: scout/utils/ensembl_biomart_clients.py [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+class EnsemblBiomartHandler:
+    """A class that handles Ensembl genes, transcripts and exons downloads via schug."""
+
+    def __init__(self, build: str = "37"):
+        self.build: str = BUILDS[build]
+
+    def biomart_get(self, url: str) -> Iterator:
+        """Sends a request to Ensembl Biomart and returns the resource lines."""
+        response: requests.models.responses = requests.get(url, stream=True)
+        return response.iter_lines(decode_unicode=True)
```
Comment (dnil) [inline scout/utils/ensembl_biomart_clients.py:29]: 👍

File: scout/utils/ensembl_biomart_clients.py [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+    def stream_resource(self, interval_type: str) -> Iterator[str]:
+        """Fetches genes, transcripts or exons from a remote Ensembl biomart in the right genome build and saves them to file."""
+
+        def yield_resource_lines(iterable) -> str:
```
Comment (dnil) [inline scout/utils/ensembl_biomart_clients.py:34]: Hm, isn't this just 
`iterable.pop(); return iterable` - or `iterable.pop(-1)` if it was on the other end?

File: tests/utils/test_ensembl_biomart_cients.py [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    (Build.build_38, TRANSCRIPTS_38_FILE_PATH),
+]
+
+SCHUG_BUILD_EXONS_PATHS = [
+    (Build.build_37, EXONS_37_FILE_PATH),
+    (Build.build_38, EXONS_38_FILE_PATH),
+]
+
+
+@pytest.mark.parametrize("build, path", SCHUG_BUILD_GENES_PATHS)
```
Comment (dnil) [inline tests/utils/test_ensembl_biomart_cients.py:33]: 🥇

**Review by dnil — APPROVED** [review]
Nice! Clean and potentially very generalisable for the other projects in the archipelago. Works like a charm locally, only minor suggestions e.g. a possible reimplementation of `.pop()` 😊!

**Comment by dnil** [comment]
`scout download everything` with main and this branch yields identical results on my local:
```
(scout39n) ➜  tmp diff -s scout_update_everything_main scout_update_everything_schug
Files scout_update_everything_main/ensembl_genes_37.txt and scout_update_everything_schug/ensembl_genes_37.txt are identical
Files scout_update_everything_main/ensembl_genes_38.txt and scout_update_everything_schug/ensembl_genes_38.txt are identical
Files scout_update_everything_main/ensembl_transcripts_37.txt and scout_update_everything_schug/ensembl_transcripts_37.txt are identical
Files scout_update_everything_main/ensembl_transcripts_38.txt and scout_update_everything_schug/ensembl_transcripts_38.txt are identical
Files scout_update_everything_main/fordist_cleaned_exac_r03_march16_z_pli_rec_null_data.txt and scout_update_everything_schug/fordist_cleaned_exac_r03_march16_z_pli_rec_null_data.txt are identical
Files scout_update_everything_main/genemap2.txt and scout_update_everything_schug/genemap2.txt are identical
Files scout_update_everything_main/genes_to_phenotype.txt and scout_update_everything_schug/genes_to_phenotype.txt are identical
Files scout_update_everything_main/hgnc.txt and scout_update_everything_schug/hgnc.txt are identical
Files scout_update_everything_main/hpo.obo and scout_update_everything_schug/hpo.obo are identical
Files scout_update_everything_main/mim2genes.txt and scout_update_everything_schug/mim2genes.txt are identical
Files scout_update_everything_main/orphadata_en_product4.xml and scout_update_everything_schug/orphadata_en_product4.xml are identical
Files scout_update_everything_main/orphadata_en_product6.xml and scout_update_everything_schug/orphadata_en_product6.xml are identical
Files scout_update_everything_main/phenotype.hpoa and scout_update_everything_schug/phenotype.hpoa are identical
Files scout_update_everything_main/phenotype_to_genes.txt and scout_update_everything_schug/phenotype_to_genes.txt are identical
```

File: scout/utils/ensembl_biomart_clients.py (hunk 10 above)
Comment (northwestwitch) [reply scout/utils/ensembl_biomart_clients.py:34]: Didn't know it existed! I'll give it a try!

Comment (dnil) [reply scout/utils/ensembl_biomart_clients.py:34]: Ah, ok; I've learned something! All iterables are not alike, and while many of them have a pop() implementation, it is presumably not in the base class. I suppose one could either just instantiate as list or other pop-able iterator, or use "generator comprehension", but all good - this works, I was mostly curios!

File: CHANGELOG.md [hunk 12]
```diff
@@ -14,6 +14,7 @@ About changelog [here](https://keepachangelog.com/en/1.0.0/)
 - CLI scout update individual accepts subject_id
 - Update ClinVar inheritance models to reflect changes in ClinVar submission API
 - Handle variant-associated condition ID format in background when creating ClinVar submissions
+- Replace the code that downloads Ensembl genes, transcripts and exons with the Schug library
```
Comment (dnil) [inline CHANGELOG.md:17]: 👍

File: scout/utils/ensembl_biomart_clients.py (hunk 10 above)
Comment (dnil) [reply scout/utils/ensembl_biomart_clients.py:34]: Absolutely - this is not wrong! One can do it differently, but pop that I suggested was not available!

**Pull request merged**